### PR TITLE
[clang-cl][flang][dxc] Fix opts exposed to clang-cl/dxc by mistake

### PR DESCRIPTION
**Pull request by Maetveis**
When these options were enabled for flang the visibility was also extended to clang-cl and dxc. I believe this was due to a misunderstanding of the default value for `Visibility`.

This is marked as draft because I would like to discuss if any of these options should remain, and once that is done add tests to ensure they or are not supported for the relevant driver modes.

### CODE REVIEW

**Comment by Maetveis** [comment]
@tarunprabhu Continuing [our discussion on #108868](https://github.com/llvm/llvm-project/pull/108868/files/612b93eda5695306293bb8ca6e27a810b06adf3d#r1867949328) here. These are what I found, by searching for commits that added both `FlangOption` and `CLOption` to Options.td and were still reachable in the blame of the file. It is possible that this is not all of them.

The default can get quite confusing to track because IIUC the `OptionGroup`'s `Visibility` gets unioned with the visibility defined on the record.

**Comment by Maetveis** [comment]
@HaohaiWen can you check if this makes sense for clang-cl? @bob80905 same for DXC please.

**Comment by tarunprabhu** [comment]
Thanks for this. I think you have caught all everything I changed. But I'll take another look to see if there was anything else.

**Comment by tarunprabhu** [comment]
I checked and I don't think there are any others that have been accidentally enabled. Are there any tests that check to these options produce warnings? If not, it might be worth adding them. What do you think?

File: clang/include/clang/Driver/Options.td
Comment (HaohaiWen) [reply clang/include/clang/Driver/Options.td:1058]: For clang-cl, /link would pass all args after it to linker and that would cause some trouble when using cmake.
e.g. cmake may use clang-cl as "linker" and append some args after flags specified by CMAKE_EXE_LINKER_FLAGS.
In that case, -Xlinker is better choice.

**Comment by HaohaiWen** [comment]
Better to double check with author who specified those CL flags.

**Comment by Maetveis** [comment]
> Better to double check with author who specified those CLOption.

Do you mean the person that added `CLOption` to these flags? That was @tarunprabhu, but IIUC that was by mistake; his intention was to extend these flags only to Flang.

Otherwise `-Xlinker` has been around for very long. I guess I will revert these flags to be clang and flang only, then open issues for the discussion of potentially enabling them to clang-cl.

**Comment by tarunprabhu** [comment]
> > Better to double check with author who specified those CLOption.
> 
> Do you mean the person that added `CLOption` to these flags? That was @tarunprabhu, but IIUC that was by mistake; his intention was to extend these flags only to Flang.

That is correct, the intention was to extend those flags only to Flang

**Comment by HaohaiWen** [comment]
> > Better to double check with author who specified those CLOption.
> 
> Do you mean the person that added `CLOption` to these flags? That was @tarunprabhu, but IIUC that was by mistake; his intention was to extend these flags only to Flang.
> 
> Otherwise `-Xlinker` has been around for very long. I guess I will revert these flags to be clang and flang only, then open issues for the discussion of potentially enabling them to clang-cl.

Some of our project have already used -Xlinker on Windows. Is there any equivalent flag for CL?
Can we keep this flag for CL in this PR and open an issue to discuss whether it should be disabled for CL?

**Comment by Maetveis** [comment]
> Can we keep this flag for CL in this PR and open an issue to discuss whether it should be disabled for CL?

Sure, see latest commit and https://github.com/llvm/llvm-project/issues/119179.

**Comment by Maetveis** [comment]
@HaohaiWen can you review and approve if it looks good now for clang-cl? I'll try pinging discord and waiting a little bit for HLSL people before merging.

Gentle ping @bob80905.

**Review by HaohaiWen — APPROVED** [review]
LGTM for -Xlinker.

**Comment by HaohaiWen** [comment]
@MaskRay, may need to merge https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/fprofile-sample-use.c to clang/test/Driver/unknown-arg-drivermodes.test

File: clang/test/Driver/unknown-arg-drivermodes.test
Comment (MaskRay) [inline clang/test/Driver/unknown-arg-drivermodes.test:13]: need --implicit-check-not=warning:

**Review by bob80905 — APPROVED** [review]
LGTM

**Comment by Maetveis** [comment]
ping @MaskRay

**Review by tarunprabhu — APPROVED** [review]
Thanks for adding the test as well. LGTM.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**
